✨ Add JSON Schema autogeneration for Pipeline YAML files

tiangolo opened the pull request:
**Proposed changes**:
- ✨ Add JSON Schema autogeneration for Pipeline YAML files

**Status (please check what you already did)**:
- [x] First draft (up for discussions & feedback)
- [ ] Final code
- [ ] Added tests
- [ ] Updated documentation

## How to use it - final users

### VS Code

As a final user wanting to test this out, open VS Code. Install the main extension to provide support for YAML (it's the most popular one and the one VS Code would probably automatically recommend installing): https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml

Then create a pipeline YAML file, for example `my-app.haystack-pipeline.yml` (the name doesn't matter *yet* at this stage).

Add this comment on top of the YAML file:

```
# yaml-language-server: $schema=https://raw.githubusercontent.com/tiangolo/haystack/json-schema/json-schemas/haystack-pipeline.0.7.schema.json
```

**Note**: this will only be needed in this early stage while testing it all manually.

Now trigger autocompletion to fill the configuration (<kbd>Ctrl</kbd> + <kbd>Space</kbd> or <kbd>Command</kbd> + <kbd>Space</kbd>).

#### Editor support - VS Code

The editor will help you fill in the details.

* Including the available components/nodes:

![Selection_080](https://user-images.githubusercontent.com/1326112/150045974-e3c72ee2-182c-4ceb-bed5-89a9a94e3177.png)

* And including their parameters with their correct types, etc:

![Selection_081](https://user-images.githubusercontent.com/1326112/150046176-aa2b977f-9d40-4a40-a2f0-800ac78c13ef.png)

* It will also show inline errors for bad types:

![Selection_082](https://user-images.githubusercontent.com/1326112/150046271-645752fe-683c-45d5-85c6-4f30b79a3cee.png)

* And for incomplete data:

![Selection_083](https://user-images.githubusercontent.com/1326112/150046305-9fae77bb-bbc6-447a-baf7-92cca2ade6ac.png)

### PyCharm

As a final user wanting to test this out, open PyCharm.

Then create a pipeline YAML file, for example `my-app.haystack-pipeline.yml` (the name doesn't matter *yet* at this stage).

Follow the [PyCharm instructions here](https://www.jetbrains.com/help/pycharm/json.html#ws_json_schema_add_custom) to associate this YAML file with the JSON Schema URL (use JSON Schema version 7):

```
https://raw.githubusercontent.com/tiangolo/haystack/json-schema/json-schemas/haystack-pipeline.0.7.schema.json
```

Alternative to those instructions, there's a small indicator in the bottom right corner that says `No JSON Schema`, you can click it and then click the `New schema mapping` option right there.

![Selection_084](https://user-images.githubusercontent.com/1326112/150046694-4c79bc2f-f5b0-460e-88cd-9536f86cbfc0.png)

**Note**: Use JSON Schema version 7:

![Selection_085](https://user-images.githubusercontent.com/1326112/150046820-f2e09a13-395f-4dda-a3be-7c9153aa2dde.png)

**Note**: all these configurations will only be needed in this early stage while testing it all manually.

#### Editor support - PyCharm

The editor will help you fill in the details.

* Including the available components/nodes:

![Selection_086](https://user-images.githubusercontent.com/1326112/150046959-763adc50-5a65-4782-a8e6-844b6bdb3949.png)

* And including their parameters:

![Selection_087](https://user-images.githubusercontent.com/1326112/150047062-0034d3a9-115b-4345-95bc-9f74296e90dc.png)

* With their correct types, etc:

![Selection_088](https://user-images.githubusercontent.com/1326112/150047158-86f56d92-108a-4383-b367-3e9c363f8cb8.png)

* It will also show inline errors:

![Selection_089](https://user-images.githubusercontent.com/1326112/150047405-26e3bc5b-8721-4d8d-9c9d-fd4105afacb1.png)

## How it works

This is all based on JSON Schema, a JSON file defining the schema (all the valid names of fields, data types, internal data shapes, etc) for the YAML file.

JSON Schema could be written by hand, but that would be quite cumbersome as it's a lot of lines and it would be difficult to keep in sync with the code.

Instead of that, there's a script `generate_json_schema.py`, it uses a lot of internal tricks and utils from Pydantic and FastAPI, plus some things and patches on top to extract the information from the parameters in the `__init__` methods (the constructors) of the Haystack node classes.

As this depends on the type annotations, this PR also updates the type annotations in a couple of places to make the information extractable by the script and the utility functions.

## GitHub Action

To make it easier to keep in sync, there also a GitHub Action that runs the script and automatically creates a PR with the new JSON Schema files.

It currently runs on "workflow dispatch", that means that you go to the GitHub Actions tab, then go to "Generate JSON Schema" and click "Run workflow":

![Selection_090](https://user-images.githubusercontent.com/1326112/150048137-f5770d57-938b-4b44-b118-5d573010d4c3.png)

**Note**: it will only work after the GitHub Action is merged to the main branch `master`, that's why I'm showing it and running it in my fork for now. 🤓

The GitHub action definition would probably be updated later to be run every time there's a new PR merged to master, or something similar.

## JSON Schema Store

The editors (and extension) support by default the schemas available in this centralized JSON Schema Store: https://www.schemastore.org/json/

Each particular schema is selected based on file name patterns.

It's all open source on GitHub: https://github.com/schemastore/schemastore/

There's documentation about how to add schemas there, and there's an option to self-host the schemas: https://github.com/SchemaStore/schemastore/blob/master/CONTRIBUTING.md#self-hosting-schemas

This option would be ideal because it allows updating the JSON Schema in Haystack's repo, and servers.

## Next steps

- The first step would be to verify that the JSON Schema is sensible, by editing pipeline files and checking the autocompletion, errors, etc.

- The next is to check if all the nodes that should be included are being included or not.

Currently it includes the classes in `haystack.document_stores` and `haystack.nodes`, filtering by the ones that are subclasses of `haystack.nodes.BaseComponent`. But it might be necessary to adjust that, remove some of those components, add more, or even create some list in the main code that includes all the nodes that should be available for the pipeline (and that should be generated in the JSON Schema).

- The next step is to decide what would be the appropriate way to handle versions.

Currently it's just fixed to `0.7`, but at which point should it increase the version? Should previous versions still be supported with older versions of the JSON Schema files? That would mean a lot of extra logic needed for the automatization in the GitHub action, to check what is the current schema version, what file to generate or update and to make the PR, etc.

Depending on this versioning scheme it would probably affect the next step.

- Define the file format/pattern to use for the pipelines.

This is because the schemas from the JSON Schema Store depend on that. for example  `*.haystack-pipeline.yml`.

- Define the URL that should store the JSON Schemas, probably something like: `https://haystack.deepset.ai/json-schemas/haystack-pipeline.0.7.schema.json`

- Make the PR(s) to JSON Schema Store including the definition for Haystack's pipeline JSON Schema.

tholor commented:
Amazing work @tiangolo - thx a lot!

We did some quick testing in PyCharm and VS Code - works like a charm :grin: .

Here are a few observations of things that didn't work as expected yet: 

> The first step would be to verify that the JSON Schema is sensible, by editing pipeline files and checking the autocompletion, errors, etc.
- Params are not "scoped" by the component type (e.g. for `DensePassageRetriever` we don't show an error for the param `api_key`  although this is not a valid param for the class - it's only a valid param for a different component). While the autocomplete in VS Code seems to work properly here (not showing `api_key`), PyCharm includes all params in a long drop-down.
![image](https://user-images.githubusercontent.com/1563902/150142965-82e130d1-5016-4e6b-9f3f-2cf882601739.png)

- No validation error for wrong `params` keys
![wrong param](https://user-images.githubusercontent.com/1563902/150138787-6ea8abaa-8313-493f-99e4-d09af81b8dfe.png)


- No validation error for wrong elements in `components`:
![image](https://user-images.githubusercontent.com/1563902/150139171-1e53eecc-94b5-4f06-9e04-243a3897f21a.png)

 - Similar for the elements in `pipelines`:
![wrong_key_for_nodes](https://user-images.githubusercontent.com/1563902/150139054-6d9ad964-a80f-4c6e-9131-55684a77c37f.png)


> The next is to check if all the nodes that should be included are being included or not.

- Base classes should not be included as valid components in the schema (BaseRanker, BaseComponent ...). Despite that the list seems complete here.


---------

> The next step is to decide what would be the appropriate way to handle versions.

We will check out a few options here (together with the naming pattern for the files), but might be something that we tackle from our side after the main work of the PR is done.

ZanSara commented:
Hi @tiangolo, cool work! Super-fast too! :slightly_smiling_face: 

I tested this branch on VSCode and I also observed most of the issues that @tholor already listed above, plus another with document stores. Apparently I can't get autocomplete for document stores consistently: sometimes it just won't be available as an option in the dropdown.

I attach a small gif that shows what works well and what doesn't:

![schema](https://user-images.githubusercontent.com/5703634/150156621-9f653d69-b7cc-47b7-bb45-e869d6ef9d7f.gif)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/deepset-ai/haystack?pullRequest=2020) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/deepset-ai/haystack?pullRequest=2020) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: tiangolo<br/>:x: github-actions[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/deepset-ai/haystack?pullRequest=2020) it.</sub>

tiangolo commented:
Thanks for the feedback @tholor and @ZanSara! I'm glad you're liking it so far. 😁

## Invalid properties

About extra parameters not being marked as invalid (this is apart from autocompletion), I just updated it to disallow additional parameters.

Now if you add `api_key` to `DensePassageRetriever` it will show an error saying that it's not a valid parameter, it seems to work well in both VS Code and PyCharm. I did the same for all the other places that could take additional properties, to make them all invalid (in `params`, `components`, `pipelines`, the root level).

Now, about autocompletion, PyCharm (IDEA) has a bug and is autocompleting invalid options. It's tracked in these two issues:

https://youtrack.jetbrains.com/issue/IDEA-274411
https://youtrack.jetbrains.com/issue/IDEA-274952

On the other hand, a more drastic approach to overcome PyCharm's bug is to alter the structure of the YAML file, I made a new PR with the alternative implementation and JSON Schema, and it seems to fix the autcompletion issue in PyCharm: https://github.com/deepset-ai/haystack/pull/2046

## Base classes

Sure! I added something to check that the name doesn't start with "Base". It will work for now, but still seems a bit fragile. 🤔

Another option to consider is having a list with all the valid node classes in the main codebase, and then the script could just use that.

In any case, the current trick should work for now. I left a TODO comment in case you want to review that later.

## Version autocompletion

I notices that both PyCharm and VS Code autocomplete the version without quotes, which is an error.

I suspect it's because YAML supports unquoted strings. But then as it starts with a `0`, it's interpreted as a number, which is invalid. But it seems like a bug in both VS Code and PyCharm including the string directly without checking that in YAML it will mean something else.

For example, if you use the same JSON Schema for a JSON file:

```JSON
{
  "$schema": "json-schemas/haystack-pipeline.0.7.schema.json",
  "version": "0.7"
}
```

...it autocompletes the version correctly, as a string.

I guess this would deserve a bug report in both InteliJ IDEA (PyCharm) and VS Code.

A "dirty" hack could be to make the version be `v0.7`. With that it would autocomplete correctly the string. But then the rest of the code would have to be able to read that version string with the prefix `v`.

## Autocompletion for Document stores (internal names)

@ZanSara I'm glad you liked it! :blush: 

I'm not sure I got the issue about Document Stores you are describing completely well.

I think it's that it doesn't autocomplete the name `MyES` in the `document_store`, right?

The problem is that that name only exists in the same YAML file, the name `ElasticSearchDocumentStore` is autocompleted for the `type` because it is defined as a constant in the JSON Schema, extracted from the class name.

But the name of that component, `MyES`, is not available anywhere in the JSON Schema, so there's no way to reference it as an option in `document_store`.

If the `document_store` required taking only preexisting/predefined classes (as is the case for `type`), it could have autocompletion, but as it can reference any name in the previous components, there's no way to define that in JSON Schema (at least not that I know of).

ZanSara commented:
Hello @tiangolo, I tried out the new version on VSCode :slightly_smiling_face: 

1. Incorrect parameter names are now detected correctly into components!
![2022-01-25_10:40:35](https://user-images.githubusercontent.com/5703634/150951816-98bf7806-3877-4907-8868-c85c90210acc.png)

2. But not outside of them. Do you think that would be possible too?
![2022-01-25_10:40:30](https://user-images.githubusercontent.com/5703634/150952068-7fd3cddd-3f8c-4a07-b016-3a126ec34c7f.png)

3. I still see the base classes into the recommendations unfortunately
![2022-01-25_10:43:56](https://user-images.githubusercontent.com/5703634/150952408-5a5e69e1-9b66-4cad-9366-52c1eb10d7ac.png)
![2022-01-25_10:44:10](https://user-images.githubusercontent.com/5703634/150952619-8e6f5ac3-b129-49e2-8d6f-bb7d1264c3d6.png)

4. Some components are missing from the suggestions. I see this issue with the document stores and the readers, while retrievers for example are suggested correctly. Here are some examples: I didn't do exhaustive testing of which components are missing yet.
 - Document Stores: missing
![2022-01-25_10:47:09](https://user-images.githubusercontent.com/5703634/150952888-cd5db8b2-b40b-499d-899c-6d45301db65a.png)
 - Readers: missing
![2022-01-25_10:49:11](https://user-images.githubusercontent.com/5703634/150953198-47faa68b-bfa2-41c9-bb23-5e6d92dedfa1.png)
 - Retrievers: found
![2022-01-25_10:51:32](https://user-images.githubusercontent.com/5703634/150953603-b5b6caef-be25-4b51-8ee7-bf4d681d4299.png)
Note that this happens only for whole components: if I explicitly enter `type:` and ask for suggestions, it works well
![2022-01-25_10:53:59](https://user-images.githubusercontent.com/5703634/150954088-032b7660-419a-4339-8723-06b0bbbcf6a4.png)

Let me know if any of this behavior is unfixable or intended, or if I have any visible issue with my setup. It's already quite nice to use!

tiangolo commented:
Thanks for the review @ZanSara! :bow: 

Has it ever happened to you that after checking for some time why some device or kitchen appliance is broken, you discover it was disconnected all along? It just happened to me. I didn't push the last commits in my branch when I switched to the other alternative implementation. :facepalm: :laughing: :facepalm: 

---

> Incorrect parameter names are now detected correctly into components!

Nice! :tada: 

> But not outside of them. Do you think that would be possible too?

This is one of the things I forgot to push. :facepalm: 

Just so you know, and for future reference for whenever you want to add more of these things in the future, this is achieved by adding to JSON Schema the property:

```
"additionalProperties": False
```

Here's the commit: https://github.com/tiangolo/haystack/commit/3973c59c3071c256ad21fe4a09fc54c7ce8b9793

> I still see the base classes into the recommendations unfortunately

Yes, I forgot to push it. :facepalm: 

> Some components are missing from the suggestions. I see this issue with the document stores and the readers, while retrievers for example are suggested correctly. Here are some examples: I didn't do exhaustive testing of which components are missing yet.

> Document Stores: missing
> Readers: missing
> Retrievers: found
> Note that this happens only for whole components: if I explicitly enter type: and ask for suggestions, it works well
> Let me know if any of this behavior is unfixable or intended, or if I have any visible issue with my setup.

Yes, in this case, we can't do anything, it's probably a limit on the number of things in the dropdown shown by VS Code and the YAML extension. :pensive: 

The good news is that if it works correctly when completing the `type`, then the JSON Schema is correct.

And to be fair, I didn't really expect VS Code to have that autocompletion of the whole component, I expected only the `type` part. :sweat_smile: 

> It's already quite nice to use!

That's so good to hear! :tada: :rocket:

ZanSara commented:
@tiangolo That's great news after all! :smile: I'll test it out again and let you know shortly :+1:

ZanSara commented:
Ok, I pulled the latest commits and I confirm it all works nicely in VSCode now! :star_struck: 

For me this looks ready, I'll approve shortly. @tholor what do you think?

ZanSara commented on .github/utils/generate_json_schema.py:
```diff
@@ -0,0 +1,256 @@
+    for module, node in possible_nodes:
+        if lenient_issubclass(
+            node, haystack.nodes.BaseComponent
+        ) and not node.__name__.startswith("Base"):
```
Note for a future refactoring: we should probably make our base classes subclasses of `abc.ABC` (https://docs.python.org/3/library/abc.html) to properly highlight the fact that they should not be used directly.

This should not be part of this PR, however I can imagine that the implementation could look like:
```python
if lenient_issubclass(node, BaseComponent) and not issubclass(node, abc.ABC):
```

ZanSara reviewed (approved):
Nice work! :slightly_smiling_face:

tiangolo commented:
Woohoo! Awesome! 🤓🚀
